ilc crashes on invalid Get

### Description

Calling `Type.GetType("System.Collections.Generic.Dictionary`2[System.String]", true);` with the incorrect number of generic arguments leads to ilc crashing. 

The crash makes it really hard to figure out what the problematic code is.  

### Reproduction Steps

1) download the attached project
2) dotnet publish

### Expected behavior


[ConsoleApp47.zip](https://github.com/dotnet/runtime/files/14348357/ConsoleApp47.zip)

```
PS C:\UnitySrc\dev\TestGround\ConsoleApp47\ConsoleApp47> dotnet run
Unhandled exception. System.ArgumentException: The number of generic arguments provided doesn't equal the arity of the generic type definition. (Parameter 'instantiation')

   at System.RuntimeType.MakeGenericType(Type[] instantiation)
   at System.Reflection.TypeNameParser.Parse()
   at System.Reflection.TypeNameParser.GetType(String typeName, Func`2 assemblyResolver, Func`4 typeResolver, Assembly requestingAssembly, Boolean throwOnError, Boolean ignoreCase, Boolean extensibleParser)
   at System.Type.GetType(String typeName, Boolean throwOnError)
   at Program.<Main>$(String[] args) in C:\UnitySrc\dev\TestGround\ConsoleApp47\ConsoleApp47\Program.cs:line 3
```

### Actual behavior

```
PS C:\UnitySrc\dev\TestGround\ConsoleApp47\ConsoleApp47> dotnet publish
MSBuild version 17.8.3+195e7f5a3 for .NET
  Determining projects to restore...
  Restored C:\UnitySrc\dev\TestGround\ConsoleApp47\ConsoleApp47\ConsoleApp47.csproj (in 227 ms).
  ConsoleApp47 -> C:\UnitySrc\dev\TestGround\ConsoleApp47\ConsoleApp47\bin\Release\net8.0\win-x64\ConsoleApp47.dll
  Generating native code
EXEC : error : Index was outside the bounds of the array. [C:\UnitySrc\dev\TestGround\ConsoleApp47\ConsoleApp47\ConsoleApp47.csproj]
  System.IndexOutOfRangeException: Index was outside the bounds of the array.
     at Internal.Runtime.CompilerHelpers.ThrowHelpers.ThrowIndexOutOfRangeException() + 0x24
     at Internal.TypeSystem.SignatureTypeVariable.InstantiateSignature(Instantiation, Instantiation) + 0x28
     at Internal.TypeSystem.InstantiatedType.InstantiateSignature(Instantiation, Instantiation) + 0x63
     at Internal.TypeSystem.InstantiatedType.InstantiateSignature(Instantiation, Instantiation) + 0x63
     at Internal.TypeSystem.InstantiatedType.InstantiateTypeArray[T](T[], Instantiation, Instantiation) + 0x4f
     at Internal.TypeSystem.MetadataRuntimeInterfacesAlgorithm.ComputeRuntimeInterfacesForInstantiatedType(InstantiatedType) + 0x54
     at Internal.TypeSystem.TypeDesc.InitializeRuntimeInterfaces() + 0x3c
     at ILCompiler.DependencyAnalysis.EETypeNode.AnalyzeVirtualMethods(TypeDesc) + 0x11a
     at ILCompiler.DependencyAnalysis.EETypeNode..ctor(NodeFactory, TypeDesc) + 0x11a
     at ILCompiler.DependencyAnalysis.NodeFactory.CreateConstructedTypeNode(TypeDesc) + 0x75
     at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey) + 0x11
     at ILCompiler.DependencyAnalysis.ReflectedTypeNode.GetStaticDependencies(NodeFactory) + 0x94
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1) + 0x44
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependencies(DependencyNodeCore`1) + 0x2b
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack() + 0xb5
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() + 0x50
     at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan() + 0x14
     at ILCompiler.Program.<Run>g__RunScanner|4_0(Program.<>c__DisplayClass4_0&) + 0x191
     at ILCompiler.Program.Run() + 0x2444
     at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass221_0.<.ctor>b__0(ParseResult result) + 0x315
C:\Users\Mike\.nuget\packages\microsoft.dotnet.ilcompiler\8.0.1\build\Microsoft.NETCore.Native.targets(305,5): error MSB3073: The command ""C:\Users\Mike\.nuget\packages\runtime.win-x64.microsoft.dotnet.ilcompiler\8.0.1\tools\\ilc" @"obj\Release\net8.0\win-x64\native\Cons 
oleApp47.ilc.rsp"" exited with code 1. [C:\UnitySrc\dev\TestGround\ConsoleApp47\ConsoleApp47\ConsoleApp47.csproj]
```

### Regression?

I don't think it's a regression from net7 nativeaot.

.NET Framework, mono and .NET Core all throw an exception at runtime

### Known Workarounds

_No response_

### Configuration

net8.  Windows x64.

### Other information

_No response_

--- a/src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Validation.cs
+++ b/src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Validation.cs
@@ -326,6 +326,12 @@ namespace ILCompiler
                     return type;
                 }
 
+                // Make sure instantiation length matches the expectation
+                if (type.Instantiation.Length != type.GetTypeDefinition().Instantiation.Length)
+                {
+                    ThrowHelper.ThrowTypeLoadException(ExceptionStringID.ClassLoadGeneral, type);
+                }
+
                 // We need to be able to load interfaces
                 foreach (var intf in type.RuntimeInterfaces)
                 {
@@ -342,12 +348,6 @@ namespace ILCompiler
                 // Ensure we can compute the type layout
                 defType.ComputeInstanceLayout(InstanceLayoutKind.TypeAndFields);
                 defType.ComputeStaticFieldLayout(StaticLayoutKind.StaticRegionSizesAndFields);
-
-                // Make sure instantiation length matches the expectation
-                if (defType.Instantiation.Length != defType.GetTypeDefinition().Instantiation.Length)
-                {
-                    ThrowHelper.ThrowTypeLoadException(ExceptionStringID.ClassLoadGeneral, type);
-                }
 
                 foreach (TypeDesc typeArg in defType.Instantiation)
                 {

--- a/src/tests/nativeaot/SmokeTests/Reflection/Reflection.cs
+++ b/src/tests/nativeaot/SmokeTests/Reflection/Reflection.cs
@@ -1509,6 +1509,7 @@ internal static class ReflectionTest
             try
             {
                 Type.GetType("System.Span`1[[System.Byte, System.Runtime]][], System.Runtime");
+                Type.GetType("System.Collections.Generic.Dictionary`2[System.String]");
             }
             catch { }
 

